Jenkinsfile

```diff
@@ -1,48 +1,67 @@
 pipeline {
-    agent any 
+    agent any
     stages {
-	stage('Install dependencies and download data') {
-	    steps {
-		echo "Dependencies"
-		sh """
-        source /var/lib/jenkins/py3env/bin/activate
-        env
 
-        cd ${env.WORKSPACE}
-        pip install -r requirements.txt
-
-        echo ${env.JOB_NAME}
-        rm -rf /tmp/${env.JOB_NAME}
-        mkdir -p /tmp/${env.JOB_NAME}/sample0
-        cd /tmp/${env.JOB_NAME}
-        git clone https://github.com/ucgmsim/qcore.git
-
-        ln -s $HOME/data/testing/slurm_gm_workflow/SGMW /tmp/${env.JOB_NAME}/build
-        cd /tmp/${env.JOB_NAME}/sample0
-        cp -r $HOME/data/testing/slurm_gm_workflow/PangopangoF29_HYP01-10_S1244/* .
+        stage('Settin up env') {
+            steps {
+                echo "[[ Start virtual environment ]]"
+                sh """
+                    echo "[ Current directory ] : " `pwd`
+                    echo "[ Environment Variables ] "
+                    env
+# Each stage needs custom setting done again. By default /bin/python is used.
+                    source /var/lib/jenkins/py3env/bin/activate
+                    mkdir -p /tmp/${env.JOB_NAME}/${env.ghprbActualCommit}
+# I don't know how to create a variable within Jenkinsfile (please let me know)
+#                   export virtenv=/tmp/${env.JOB_NAME}/${env.ghprbActualCommit}/venv
+                    python -m venv /tmp/${env.JOB_NAME}/${env.ghprbActualCommit}/venv
+# activate new virtual env
+                    source /tmp/${env.JOB_NAME}/${env.ghprbActualCommit}/venv/bin/activate
+                    echo "[ Python used ] : " `which python`
+                    cd ${env.WORKSPACE}
+                    echo "[ Install dependencies ]"
+                    pip install -r requirements.txt
+                    echo "[ Install qcore ]"
+                    cd /tmp/${env.JOB_NAME}/${env.ghprbActualCommit}
+                    rm -rf qcore
```

Comment:
are these lines only used to have a up-to-date qcore?
just wondering why using rm and clone, instead of just git pull.
because after the most recent changes that added data to the repo, a full clone is going to take a lot longer than just a pull

Reply:
rm is done "just-in-case". In most cases, if it is cloned at /tmp/qcore/{env.ghprbActualCommit}, qcore should be empty. Sometimes, env.ghprbActualCommit can be a branch name (especially when manually triggered via Jenkins web interface), and in such a case qcore may not be empty. It's safer to delete whatever is there and pull the fresh clone.